raft.go

```diff
@@ -310,7 +323,41 @@ func (r *Raft) runCandidate() {
 		case rpc := <-r.rpcCh:
 			r.mainThreadSaturation.working()
 			r.processRPC(rpc)
+		case preVote := <-prevoteCh:
+			// This a pre-vote case it should trigger a "real" election if the pre-vote is won.
+			r.mainThreadSaturation.working()
+			r.logger.Debug("pre-vote received", "from", preVote.voterID, "term", preVote.Term, "tally", preVoteGrantedVotes)
+			// Check if the term is greater than ours, bail
+			if preVote.Term > term {
+				r.logger.Debug("pre-vote denied: found newer term, falling back to follower", "term", preVote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(preVote.Term)
+				return
+			}
 
+			// Check if the preVote is granted
+			if preVote.Granted {
+				preVoteGrantedVotes++
+				r.logger.Debug("pre-vote granted", "from", preVote.voterID, "term", preVote.Term, "tally", preVoteGrantedVotes)
+			} else {
+				preVoteRefusedVotes++
+				r.logger.Debug("pre-vote denied", "from", preVote.voterID, "term", preVote.Term, "tally", preVoteGrantedVotes)
+			}
+
+			// Check if we've won the pre-vote and proceed to election if so
+			if preVoteGrantedVotes >= votesNeeded {
+				r.logger.Info("pre-vote successful, starting election", "term", preVote.Term, "tally", preVoteGrantedVotes, "votesNeeded", votesNeeded-1)
```

Comment:
🤔 I'm trying to work out why we `-1` the votesNeeded here. I think it's because we auto-vote for ourselves in `votesNeeded` and so it might be confusing to operators whether `votesNeeded` includes that self-vote or not.

I.e. in a cluster of 3 `votesNeeded` would actually be 2, but we effectively only need one vote from another server in addition to our own so you've chosen to show it as 1 instead here.

I initially thought this was potentially more surprising because for those who know about consensus, they'd maybe expect this number to match the quorum number of the cluster. I changed my mind when I considered the UX and assumed we'd only see "vote granted" logs for the remote peers so having this number correlate to that might be more intuitive. But then I went and looked at `preElectSelf` again and we do Debug log "voting for self".

I don't think ultimately it's a big deal but on balance it seems slightly less confusing to not subtract 1 here both to match any intuitions about quorum size, and to remain consistent with debug log output where we make explicit that we vote for ourselves too?

Reply:
At this level our own vote is completely transparent to us. It come through the `preVoteCh` like any other vote. see: https://github.com/hashicorp/raft/blob/1912201252579c050610c4a93ccfc05356647872/raft.go#L2105-L2105

Reply:
huh... then that makes me wonder even more why we have `votesNeeded-1` in the log here?

Reply:
yes you are right, I misread your comment and was thinking that it's not related to the log 😅. I fixed that and added the number of refused prevotes to the log.